### PR TITLE
[charts] Remove compatibility layer for React vs native events

### LINKED ISSUE #20957: [charts] Remove compatibility layer for React vs native events
Remove the compatibility layer for handling React and native events that was added in https://github.com/mui/mui-x/pull/20697.

Document the impact for users consuming event objects from chart callbacks or plugins. 

Parent issue: #17955

### PR DESCRIPTION
**Pull request by JCQuintas**
Try it here: https://stackblitz.com/edit/nzfsk7ey?file=src%2FDemo.tsx

## Summary

- Centralizes pointer interaction handling in `ChartsLayerContainer` instead of attaching event handlers to individual chart items (bars, lines, marks, heatmap cells)
- Removes the heatmap `shouldRegisterPointerInteractionsGlobally` compatibility layer added in #20697
- Adds `getItemAtPosition` to bar and line series configs, enabling container-level hit detection for these chart types
- Refactors `useRegisterPointerInteractions` to iterate through all series configs rather than requiring per-chart-type registration

### Breaking changes

- `BarPlot.onItemClick` now receives a native `MouseEvent` instead of `React.MouseEvent`
- `useInteractionItemProps` no longer accepts a `skip` parameter
- Heatmap custom cell slots with `onPointerEnter` in slotProps no longer fall back to item-level event handlers

Closes #20957

### CODE REVIEW

File: packages/x-charts/src/BarChart/BarChart.test.tsx [hunk 1]
```diff
@@ -149,7 +149,14 @@ describe('<BarChart />', () => {
...
 
-      await user.pointer({ target: bars[0] });
+      const barRect = bars[0].getBoundingClientRect();
+      await user.pointer({
+        target: bars[0],
+        coords: {
+          clientX: barRect.left + barRect.width / 2,
+          clientY: barRect.top + barRect.height / 2,
+        },
+      });
```
Comment (JCQuintas) [inline packages/x-charts/src/BarChart/BarChart.test.tsx:159]: `user.pointer()` creates events with clientX/clientY defaulting to (0, 0), so we need to explicitly set them now. 

This is because under the hood it doesn't fire real events, it only checks if the `target` has changed and fires the events that would supposedly happen from that.

Now we need to manually move the pointer to where we need.

File: docs/data/migration/migration-charts-v8/migration-charts-v8.md [hunk 2]
```diff
@@ -366,6 +366,29 @@ After running the codemod make sure to adapt the hook returned value to your nee
...
+
+The `skip` parameter has been removed from `useInteractionItemProps`.
+If you were using it, you can remove it — the hook now always returns interaction props.
+
+```diff
+-const interactionProps = useInteractionItemProps(identifier, skip);
++const interactionProps = useInteractionItemProps(identifier);
+```
+
+### `useRegisterPointerInteractions` signature changed
```
Comment (JCQuintas) [inline docs/data/migration/migration-charts-v8/migration-charts-v8.md]: Do we want both of these to be exported? 🤔

Comment (alexfauquette) [reply docs/data/migration/migration-charts-v8/migration-charts-v8.md]: This one is an internal hook need to build batch rendering

File: packages/x-charts/src/BarChart/checkClickEvent.test.tsx [hunk 3]
```diff
@@ -185,21 +185,32 @@ describe('BarChart - click event', () => {
 
       const rectangles = document.querySelectorAll<HTMLElement>('rect.MuiBarChart-element');
 
-      await user.click(rectangles[0]);
+      const getCenter = (el: Element) => {
+        const rect = el.getBoundingClientRect();
+        return { clientX: rect.left + rect.width / 2, clientY: rect.top + rect.height / 2 };
+      };
```
Comment (alexfauquette) [inline packages/x-charts/src/BarChart/checkClickEvent.test.tsx]: I think you can move this in the the helpers

**Comment by JCQuintas** [comment]
> yellow area

I've pushed a commit that should handle this. Seems to be working correctly as well.

**Comment by alexfauquette** [comment]
> I've pushed a commit that should handle this. Seems to be working correctly as well.

I like the effort, but it's buggy. I tried to modify the demo to test edge cases like:

- single lines
- lines with null values
- mix of line and area

Here for example the highlight swich between blue line and yellow area when pointer cross the red dot line.

image

------

IMO this feature has a grate potential. I will be much better than the current hack explained in https://mui.com/x/react-charts/line-demo/#larger-interaction-area

But it should be opt-in for some time to catch bugs and edge case while having a fallback for users.

- During v9 users get an imperfect highlight management that work for sure
- They can enable the global pointer detection to get access to a finer config, like highlighting/clicking the closest line/point
- for v10 we switch to the new interaction by default

Especially because the new solution is an approximation, whereas the current implementation is exact because it relies on the SVG geometry

I say it's an approximation because for all curves you do a linear interpolation between prev and next data point which is a first order approximation, but to get the exact value we will need more complex pipeline. For example catmul interpolation between 2 points also depends on the two points before and the two points after

image

File: docs/data/migration/migration-charts-v8/migration-charts-v8.md [hunk 4]
```diff
@@ -370,6 +370,30 @@ After running the codemod make sure to adapt the hook returned value to your nee
...
+```
+
+### Removed item-level pointer handlers
+
+Individual chart elements no longer attach `onPointerEnter`/`onPointerLeave` event handlers for highlight and tooltip interactions.
+These interactions are now handled at the container level using position-based hit detection.
+
+This affects the following components:
+
+- **Line Chart**: `MarkElement`, `CircleMarkElement`, `LineElement`, `AreaElement`
```
Comment (alexfauquette) [inline docs/data/migration/migration-charts-v8/migration-charts-v8.md]:
```suggestion
```
